pkg/controllers/workgenerator/controller.go

```diff
@@ -328,25 +381,47 @@ func (r *Reconciler) fetchAllResourceSnapshots(ctx context.Context, resourceBind
 	return resourceSnapshots, nil
 }
 
-// upsertWork creates or updates the work for the corresponding resource snapshot.
-// it returns if any change is made to the work and the possible error code.
-func (r *Reconciler) upsertWork(ctx context.Context, work *fleetv1beta1.Work, workName string, resourceSnapshot *fleetv1beta1.ClusterResourceSnapshot,
-	resourceBinding *fleetv1beta1.ClusterResourceBinding) (bool, error) {
-	needCreate := false
-	var workObj klog.ObjectRef
-	resourceBindingObj := klog.KObj(resourceBinding)
-	resourceSnapshotObj := klog.KObj(resourceSnapshot)
-	// we already checked the label in fetchAllResourceSnapShots function so no need to check again
-	resourceIndex, _ := labels.ExtractResourceIndexFromClusterResourceSnapshot(resourceSnapshot)
-	if work == nil {
-		needCreate = true
-		work = &fleetv1beta1.Work{
+// getConfigMapEnvelopWorkObj first try to locate a work object for the corresponding envelopObj of type configMap.
+// we create a new one if the work object doesn't exist. We do this to avoid repeatedly delete and create the same work object.
+// TODO: take into consider the override policy in the future
+func (r *Reconciler) getConfigMapEnvelopWorkObj(ctx context.Context, workNamePrefix string, resourceBinding *fleetv1beta1.ClusterResourceBinding,
+	resourceSnapshot *fleetv1beta1.ClusterResourceSnapshot, envelopeObj *unstructured.Unstructured) (*fleetv1beta1.Work, error) {
+	// we group all the resources in one configMap to one work
+	manifest, err := extractResFromConfigMap(envelopeObj)
+	if err != nil {
+		klog.ErrorS(err, "configMap has invalid content", "snapshot", klog.KObj(resourceSnapshot),
+			"resourceBinding", klog.KObj(resourceBinding), "configMapWrapper", klog.KObj(envelopeObj))
+		return nil, controller.NewUserError(err)
+	}
+	klog.V(2).InfoS("Successfully extract the enveloped resources from the configMap", "numOfResources", len(manifest), "configMapWrapper", klog.KObj(envelopeObj))
+	// Try to see if we already have a work represent the same enveloped object for this CRP in the same cluster
+	envelopWorkLabelMatcher := client.MatchingLabels{
```

Comment:
Hi Ryan! Is there any particular reason why the ParentResourceSnapshotIndex label is omitted for this?

Reply:
yes, this is actually critical. The work is associated with the same CRP/binding and enveloped configmap all the time. However, it can move from one snapshot to another. Sometimes the content of the configmap doesn't change between snapshots but it can also change. Thus, the ParentResourceSnapshotIndex cannot be part of the label matching. You can see that we modify the ParentResourceSnapshotIndex  and the manifest in the new work.